File: 2023/day_5/solution.js
const fs = require('fs');

function readFile(path = "input.txt") {
    const fileContent = fs.readFileSync(path, 'utf-8');
    return fileContent.split('\n').map(line => line.trim());
}

function getSeeds(input) {
    return input[0].split(":")[1].split(/\s+/);
}

function getMaps(input) {
    const maps = [];
    const mapInput = input.slice(2);
    let currentMap = [];

    for (const line of mapInput) {
        if (line === "") {
            maps.push([...currentMap]);
            currentMap = [];
        } else {
            currentMap.push(line);
        }
    }

    if (currentMap.length > 0) {
        maps.push([...currentMap]);
    }

    return maps;
}

function createConversionDictionary(map) {
    const mapping = {};
    let destinationRange, sourceRange, rangeLength, source, destination;

    for (let i = 1; i < map.length; i++) {
        [destinationRange, sourceRange, rangeLength] = convertLine(map[i]);

        for (let j = 0; j < rangeLength; j++) {
            source = sourceRange + j;
            destination = destinationRange + j;

            // Only store mappings if source and destination are different
            if (source !== destination) {
                mapping[source] = destination;
            }
        }
    }
    return mapping;
}

function convertLine(line) {
    const [destinationRange, sourceRange, rangeLength] = line.split(/\s+/).map(Number);
    return [destinationRange, sourceRange, rangeLength];
}

function runTheConversion(seed, mapArr) {
    let changingSeed = seed;
    for (const mapping of mapArr) {
        changingSeed = mapping[changingSeed] || changingSeed;
    }
    return changingSeed;
}

function solvePart1() {
    console.log("solving");
    const inputData = readFile();
    const seeds = getSeeds(inputData);
    const maps = getMaps(inputData);

    const conversionMapArr = maps.map(createConversionDictionary);

    const locationNums = seeds.map(seed => runTheConversion(parseInt(seed), conversionMapArr));

    console.log("min:", Math.min(...locationNums));
}

solvePart1();

File: 2023/day_5/solution.py
def readFile(path="input.txt"):
    file_arr = []
    with open(path, 'r') as file:
        for line in file:
            file_arr.append(line.strip())
    return file_arr

def getSeeds(input):
    seeds = input[0].split(":")[1].split()
    return seeds

def getMaps(input):
    maps = []
    mapInput = input[2:]
    currentMap = []
    for line in mapInput:
        if line == "":
            maps.append(currentMap)
            currentMap = []
            continue
        else:
            currentMap.append(line)
    if currentMap != []:
        maps.append(currentMap)
    return maps

def createConversionDictionary(map):
    # destination, source, length
    mapping = {}
    for i in range(1,len(map)):
        destinationRange, sourceRange, rangeLength = convertLine(map[i])
        for j in range(rangeLength):
            if sourceRange+j not in mapping:
                mapping[sourceRange+j] = destinationRange + j
    # print("mapping: ", mapping)
    return mapping 


def convertLine(line):
    input = line.split()
    destinationRange = int(input[0])
    sourceRange = int(input[1])
    rangeLength = int(input[2])
    return destinationRange, sourceRange, rangeLength

def runTheConversion(seed, mapArr):
    changingSeed = seed
    for map in mapArr:
        if changingSeed in map:
            changingSeed = map[changingSeed]
    return changingSeed

def solvePart1():
    print("solving")
    input = readFile()
    seeds = getSeeds(input)
    maps = getMaps(input)
    print(maps)
    # conversionMapArr = []
    # locationNums = []
    # for map in maps:
    #     conversionMapArr.append(createConversionDictionary(map))
    # for seed in seeds:
    #     print("seed: ", seed)
    #     intSeed = int(seed)
    #     locationNums.append(runTheConversion(intSeed, conversionMapArr))
    # print("min: ",min(locationNums))
        

solvePart1()